aten/src/ATen/native/Loss.cpp

```diff
@@ -103,6 +103,59 @@ Tensor kl_div_backward_cpu(const Tensor& grad, const Tensor& input, const Tensor
   return grad_input;
 }
 
+Tensor binary_cross_entropy_cpu(const Tensor& input, const Tensor& target, const Tensor& weight, int64_t reduction) {
+    Tensor loss;
+    return at::native::binary_cross_entropy_out_cpu(loss, input, target, weight, reduction);
+}
+
+Tensor& binary_cross_entropy_out_cpu(Tensor& loss, const Tensor& input, const Tensor& target, const Tensor& weight, int64_t reduction) {
+    TORCH_CHECK(
+        (input >= 0).mul_(input <= 1).all().item<bool>(),
+        "all elements of input should be between 0 and 1"
+    );
```

Comment:
Your CPU performance is worse on small sizes and I suspect that this line is the culprit. In the original implementation, the tests for in-boundedness happened within the loop. Now you are doing a bunch of extra allocations and memory accesses to compute this `all()` test. The original implementation seems better to me.

Reply:
Oh, but I guess you're not doing a manually fused kernel anymore, so you can't this way. I guess this will have to do...

Reply:
I'm not sure what a manually fused kernel means. Is there a way to do that in ATen? If so, I could try it out.

Reply:
Yes, it would be using something like TensorIterator to do all of the operations in one go. But you got a pretty big speed up for the larger cases on CPU so this might not be worth the effort.

Reply:
I actually just figured out how to use TensorIterator, and I believe I'm getting way better performance on the CPU forward method, unless I'm deceiving myself somehow. After I double check myself, I'll go ahead and push that change.

Reply:
For the record, there's a wiki page on using TensorIterator https://github.com/pytorch/pytorch/wiki/How-to-use-TensorIterator

Reply:
Thanks @ngimel! I don't think I ever noticed that there was a wiki before.